# 4.6 Combining and Exporting Data
### Import new dataset 'orders_products_prior.csv', import 'df_ords_checked.csv' & 'df_prods_checked.csv'
### Concatenate and merge dataframes
### Export merged data using the pickle format ('orders_products_combined.pkl')

## Import data

In [35]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [36]:
path = r'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [37]:
path

'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [38]:
#import new data set 'orders_products_combined'
df_ords_prior =df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_products_prior.csv'), index_col = False)

In [39]:
#import df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_checked.csv'), index_col = False)

In [40]:
#import df_prods
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'df_prods_checked.csv'), index_col = False)

### Check dimensions of datasets

In [41]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [42]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_time  days_since_prior_order      first_order  
0                   2           8                     NaN      First Order  
1                   3           7                    15.0  Repeat Customer  
2                   3          12                    21.0  Repeat Customer  
3                   4           7                    29.0  Repeat Customer  
4                   4          15                    28.0  Repeat Customer

In [43]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

### drop unescessary columns to save space

In [48]:
df_ords = df_ords.drop(['Unnamed: 0.1', 'Unnamed: 0'], axis=1)

In [49]:
#re-check df_ords now without those columns
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order      first_order  
0                     NaN      First Order  
1                    15.0  Repeat Customer  
2                    21.0  Repeat Customer  
3                    29.0  Repeat Customer  
4                    28.0  Repeat Customer

### Check the shape (rows/columns) of all data sets

In [44]:
df_ords_prior.shape

(32434489, 4)

In [45]:
df_ords.shape

(3421083, 9)

In [46]:
df_prods.shape

(49672, 6)

### Merge datasets: df_ords with df_ords_prior (using 'order_id' as a key)

In [50]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [51]:
#check the newly merged dataframe 
#and checking for full match with indicator = true, _merge column shows 'both'
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329        1             1                   2           8   
1   2539329        1             1                   2           8   
2   2539329        1             1                   2           8   
3   2539329        1             1                   2           8   
4   2539329        1             1                   2           8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN  First Order         196                  1   
1                     NaN  First Order       14084                  2   
2                     NaN  First Order       12427                  3   
3                     NaN  First Order       26088                  4   
4                     NaN  First Order       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [53]:
# using value_counts to quickly sum up the values in the _merge column
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

### this looks like it is a full match, but it actually isn't -using outer merge would show the columns that actually match - so test merge (without over-writing), but do not worry for this project - below would be the syntax (producing 32640698 rows × 10 columns).  This does not match 32434489

#pd_ords.merge(df_ords_prior, on = ['order_id'], how = 'outer')

In [80]:
pd.merge(df_ords, df_ords_prior, on= ['order_id'], how = 'outer')

order_id  user_id  order_number  orders_day_of_week  order_time  \
0          2539329        1             1                   2           8   
1          2539329        1             1                   2           8   
2          2539329        1             1                   2           8   
3          2539329        1             1                   2           8   
4          2539329        1             1                   2           8   
...            ...      ...           ...                 ...         ...   
32640693   2977660   206209            13                   1          12   
32640694   2977660   206209            13                   1          12   
32640695   2977660   206209            13                   1          12   
32640696   2977660   206209            13                   1          12   
32640697    272231   206209            14                   6          14   

          days_since_prior_order      first_order  product_id  \
0                            NaN      First Order       196.0   
1                            NaN      First Order     14084.0   
2                            NaN      First Order     12427.0   
3                            NaN      First Order     26088.0   
4                            NaN      First Order     26405.0   
...                          ...              ...         ...   
32640693                     7.0  Repeat Customer     38730.0   
32640694                     7.0  Repeat Customer     31477.0   
32640695                     7.0  Repeat Customer      6567.0   
32640696                     7.0  Repeat Customer     22920.0   
32640697                    30.0  Repeat Customer         NaN   

          add_to_cart_order  reordered  
0                       1.0        0.0  
1                       2.0        0.0  
2                       3.0        0.0  
3                       4.0        0.0  
4                       5.0        0.0  
...                     ...        ...  
32640693                6.0        0.0  
32640694                7.0        0.0  
32640695                8.0        0.0  
32640696                9.0        0.0  
32640697                NaN        NaN  

[32640698 rows x 10 columns]

In [81]:
df_merged_large.to_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_combined.pkl'))